controllers/receiver_controller.go

```diff
@@ -18,172 +18,212 @@ package controllers
+			msg := fmt.Sprintf("Reconciliation finished in %s, next run in %s",
+				time.Since(reconcileStart).String(), obj.Spec.Interval.Duration.String())
+			log.Info(msg)
+			r.Event(obj, corev1.EventTypeNormal, meta.SucceededReason, msg)
 		}
 
-		// Finally, patch the resource
-		if err := patchHelper.Patch(ctx, receiver, patchOpts...); err != nil {
-			retErr = errors.NewAggregate([]error{retErr, err})
+		// Patch finalizers, status and conditions.
+		if err := r.patch(ctx, obj, patcher); err != nil {
+			retErr = kerrors.NewAggregate([]error{retErr, err})
 		}
-
-		// Always record readiness and duration metrics
-		r.Metrics.RecordReadiness(ctx, receiver)
-		r.Metrics.RecordDuration(ctx, receiver, start)
-
 	}()
 
-	if !controllerutil.ContainsFinalizer(receiver, v1beta1.NotificationFinalizer) {
-		controllerutil.AddFinalizer(receiver, v1beta1.NotificationFinalizer)
+	if !controllerutil.ContainsFinalizer(obj, apiv1.NotificationFinalizer) {
+		controllerutil.AddFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{Requeue: true}
 		return
 	}
 
-	if !receiver.ObjectMeta.DeletionTimestamp.IsZero() {
-		controllerutil.RemoveFinalizer(receiver, v1beta1.NotificationFinalizer)
+	if !obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		controllerutil.RemoveFinalizer(obj, apiv1.NotificationFinalizer)
 		result = ctrl.Result{}
 		return
 	}
 
-	return r.reconcile(ctx, receiver)
-}
-
-func (r *ReceiverReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return r.SetupWithManagerAndOptions(mgr, ReceiverReconcilerOptions{})
-}
+	// Return early if the object is suspended.
+	if obj.Spec.Suspend {
+		log.Info("Reconciliation is suspended for this object")
+		return ctrl.Result{}, nil
+	}
 
-func (r *ReceiverReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts ReceiverReconcilerOptions) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&v1beta1.Receiver{}).
-		WithOptions(controller.Options{
-			MaxConcurrentReconciles: opts.MaxConcurrentReconciles,
-			RateLimiter:             opts.RateLimiter,
-			RecoverPanic:            true,
-		}).
-		Complete(r)
+	return r.reconcile(ctx, obj)
 }
 
 // reconcile steps through the actual reconciliation tasks for the object, it returns early on the first step that
 // produces an error.
-func (r *ReceiverReconciler) reconcile(ctx context.Context, obj *v1beta1.Receiver) (ctrl.Result, error) {
+func (r *ReceiverReconciler) reconcile(ctx context.Context, obj *apiv1.Receiver) (ctrl.Result, error) {
 	// Mark the resource as under reconciliation
-	conditions.MarkReconciling(obj, meta.ProgressingReason, "")
+	conditions.MarkReconciling(obj, meta.ProgressingReason, "Reconciliation in progress")
 
 	token, err := r.token(ctx, obj)
 	if err != nil {
-		conditions.MarkFalse(obj, meta.ReadyCondition, v1beta1.TokenNotFoundReason, err.Error())
-		return ctrl.Result{}, err
+		conditions.MarkFalse(obj, meta.ReadyCondition, apiv1.TokenNotFoundReason, err.Error())
+		return ctrl.Result{Requeue: true}, err
```

Comment:
In case of source-controller and image-reflector, we retain the old status data even when something fails.
Does that apply here too? Should we keep the old URL in the status when token extraction fails? Unlike the data in source-controller and image-reflector, is an old URL useful here?
Since the web server will not use a receiver that's not ready, I'm unable to see any use of keeping the old URL in the status when the receiver fails.
Just trying to think through the new concerns with periodically reconciling behavior of NC objects.